src/common.c

```diff
@@ -707,60 +709,191 @@ int fi_epoll_create(struct fi_epoll **ep)
 
 int fi_epoll_add(struct fi_epoll *ep, int fd, uint32_t events, void *context)
 {
-	struct pollfd *fds;
-	void *contexts;
-
-	if (ep->nfds == ep->size) {
-		fds = calloc(ep->size + 64,
-			     sizeof(*ep->fds) + sizeof(*ep->context));
-		if (!fds)
-			return -FI_ENOMEM;
-
-		ep->size += 64;
-		contexts = fds + ep->size;
+	struct fi_epoll_work_item *item;
 
-		memcpy(fds, ep->fds, ep->nfds * sizeof(*ep->fds));
-		memcpy(contexts, ep->context, ep->nfds * sizeof(*ep->context));
-		free(ep->fds);
-		ep->fds = fds;
-		ep->context = contexts;
-	}
+	item = calloc(1,sizeof(*item));
+	if (!item)
+		return -FI_ENOMEM;
 
-	ep->fds[ep->nfds].fd = fd;
-	ep->fds[ep->nfds].events = events;
-	ep->context[ep->nfds++] = context;
+	item->fd = fd;
+	item->events = events;
+	item->context = context;
+	item->type = FI_EPOLL_ADD;
+	fastlock_acquire(&ep->lock);
+	slist_insert_tail(&item->entry, &ep->work_item_list);
 	fd_signal_set(&ep->signal);
+	fastlock_release(&ep->lock);
 	return 0;
 }
 
 int fi_epoll_mod(struct fi_epoll *ep, int fd, uint32_t events, void *context)
 {
+	struct slist_entry *entry, *prev;
+	struct fi_epoll_work_item *item, *mod_item;
 	int i;
 
-	for (i = 0; i < ep->nfds; i++) {
-		if (ep->fds[i].fd == fd) {
-			ep->fds[i].events = events;
-			ep->context[i] = context;
-			fd_signal_set(&ep->signal);
-			return 0;
-		}
+	mod_item = calloc(1,sizeof(*mod_item));
+	if (!mod_item)
+		return -FI_ENOMEM;
+
+	fastlock_acquire(&ep->lock);
+	slist_foreach(&ep->work_item_list, entry, prev) {
+		item = container_of(entry, struct fi_epoll_work_item, entry);
+		if (item->fd != fd)
+			continue;
+
+		item->events = events;
+		item->context = context;
+		fd_signal_set(&ep->signal);
+		fastlock_release(&ep->lock);
+		free(mod_item);
+		return 0;
+	}
+
+	for (i = 1; i < ep->nfds; i++) {
+		if (ep->fds[i].fd != fd)
+			continue;
+
+		mod_item->fd = fd;
+		mod_item->events = events;
+		mod_item->context = context;
+		mod_item->type = FI_EPOLL_MOD;
+		slist_insert_tail(&mod_item->entry,
+				  &ep->work_item_list);
+		fd_signal_set(&ep->signal);
+		fastlock_release(&ep->lock);
+		return 0;
   	}
+	fastlock_release(&ep->lock);
+	free(mod_item);
```

Comment:
The only reason to loop in this call is to validate the fd.  Since this is an internal abstraction, we should be able to just allocated mod_item and queue it.  The handling code can validate the fd.

Reply:
I am mainly doing this to give user feedback/error. Should we silently discard the fd in case of user programming error?

Reply:
The user in this case is one of our providers, which is a coding error.  So I'm okay silently discarding the fd in the handler code.